### PR TITLE
Improved startup performance and measurement tools.

### LINKED ISSUE #3154: Reducing startup timing (dcc module)
I'm using the dcc module (from @vsky279 ) and I'm having problems getting the nodemcu platform to boot fast enough (I think) to support service mode writes. Does it work for you?

The issue is that there is only a limited amount of time between when the service mode programmer powers up the rails before the write commands are issued and the rails powered down. [Reads work fine (mostly)]

I'm intending to try setting the cpu speed high in the user_init function, getting rid of the startup banner, and using LFS (as I suspect that is faster than loading a `.lc` file). With lua53, the gc mode is already set sensibly. Are there other speedups that anybody can think of?

### PR DESCRIPTION
**Pull request by pjsg**
Fixes #3154 (partially)

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/*`.

This adds three different startup options that allow various tradeoffs during boot. In my environment, this leads to boot only needing 60% of the time that it did before. Of course, stuff like the delay of SPIFFS mounting just shuffles the time around, but it can be useful in special cases.

### CODE REVIEW

File: app/lua53/lua.c
Comment (nwf) [inline app/lua53/lua.c:112]: Maybe slightly better to just avoid calling `print_version` in the first place?

File: app/modules/node.c
Comment (nwf) [inline app/modules/node.c:905]: Maybe `START_*` instead of `OPTION_*`?

File: app/platform/platform.h
Comment (nwf) [inline app/platform/platform.h:404]: Split large lines with `\` escapes, here and below?

File: app/user/user_main.c
Comment (pjsg) [reply app/user/user_main.c:133]: This is an interesting question -- I'm not sure how the existing switch to 160MHZ from node.setcpufreq works with respect to rtctime....

**Comment by TerryE** [comment]
@pjsg, it is valuable seeing what can be achieved here and I feel that we generally need to move forward. However, I would really prefer that we finish the #3154 discussion or in fact open a proper **Startup Options** issue and agree scope and approach before we raise the PR in its final form.  Until we've done this, I really see this as a "do not merge yet" PR.

File: app/modules/node.c
Comment (nwf) [inline app/modules/node.c:61]: Add a deprecation warning?

File: app/platform/platform.h
Comment (nwf) [inline app/platform/platform.h:358]: Since this lives nominally in our "HAL" layer (of sorts...), maybe `STARTUP_OPTION_CPU_FREQ_MAX` or something?

File: app/platform/platform.h
Comment (nwf) [inline app/platform/platform.h:359]: I might just be missing it, but I don't see the `LROT` entries I expect to see for these to be exported to Lua?

File: app/modules/node.c
Comment (pjsg) [reply app/modules/node.c:61]: I add a message as a comment and also updated the documentation

File: app/platform/platform.h
Comment (pjsg) [reply app/platform/platform.h:359]: They aren't there -- the interface was changed as a result of the discussion, so you now say 

```
node.startup({banner=false})
```

Comment (pjsg) [reply app/platform/platform.h:359]: I could add constants for node.STARTUP_BANNER = "banner" but I'm not sure that that is really necessary.

File: docs/modules/node.md
Comment (pjsg) [reply docs/modules/node.md:568]: Done

File: app/user/user_main.c
Comment (pjsg) [reply app/user/user_main.c]: I've shuffled this around as you recommended.

File: app/platform/platform.h
Comment (nwf) [reply app/platform/platform.h:359]: No, indeed, that's not necessary.  Thank you for reminding me why I don't see the LROT entries. :)

**Comment by pjsg** [comment]
The startup counts stuff is not compiled in by default -- I agree that most people wont want it. I built the sane interface to it because I like typing `=sjson.encode(node.startupcounts())` into esplorer! (`sjson.encode` is my goto way of displaying table structures)

File: app/platform/platform.c
Comment (pjsg) [reply app/platform/platform.c:989]: Indeed -- I'll make that change.

File: app/modules/node.c
Comment (pjsg) [reply app/modules/node.c:737]: Because I was a doofus.

File: app/modules/node.c
Comment (HHHartmann) [inline app/modules/node.c:119]: This write is called even if no changes happened. Maybe compare to the original value.

File: app/modules/node.c
Comment (HHHartmann) [inline app/modules/node.c:93]: Maybe only write on changed  data. 
People might just have the setup call in their startup routine rather than just calling it once

File: app/modules/node.c
Comment (pjsg) [reply app/modules/node.c:119]: Yes -- it turns out that the platform_rcr_write already checks fo 'no change' and returns quickly in that case. I didn't feel it was worth the increased code size to shave a few microseconds off this call (by duplicating the check here)

File: app/modules/node.c
Comment (pjsg) [reply app/modules/node.c:93]: As above...

**Comment by TerryE** [comment]
@pjsg Philip, we seem to be moving to consensus on this one.  I do face one challenge here and that is that I am almost done on preparing a PR which going to sweep to know issues with Lua 5.3 and cleaning up the 5.1 / 5.3 build compatibility in terms of APIs.  This encompasses #3152, #3165, #3165 (part), #3175, #3176 and #3180.  I know that @marcelstoer prefers one issue = one PR, but in this case the secondary Issues where usually thrown up as sub-points to the main tidy-up.  As you can imagine my PR also changes most if not all of the files that you have updated, and having to rebase against `dev` is going to be a real PITA for me if this goes in first.

So can I ask that we hold off merging this PR in until after mine lands?  I am quite willing to do the knock-on rebasing work rather than impose this on you, but it's just that its going to be less work for me this way.  Thanks Terry

**Comment by pjsg** [comment]
That is fine.

File: app/modules/node.c
Comment (HHHartmann) [reply app/modules/node.c:119]: Ok. Wasn't aware of that.

**Comment by TerryE** [comment]
> That is fine.

Given that we've agreed to do this after #3193, I will temporarily add the blocked label.

**Comment by marcelstoer** [comment]
As #3193 has since been merged I removed the "blocked" label.

File: docs/modules/node.md
Comment (marcelstoer) [inline docs/modules/node.md:613]: Please wrap this with an admonition like we e.g. do for `wifi.startsmart()` https://nodemcu.readthedocs.io/en/latest/modules/wifi/#wifistartsmart. I doubt we are consistent as to the admonition _type_ (note, important, etc.).

Comment (pjsg) [reply docs/modules/node.md]: Fixed.

**Comment by pjsg** [comment]
@HHHartmann @nwf @TerryE I think that I have addressed all of your comments....

**Review by nwf — APPROVED** [review]

File: docs/modules/node.md
Comment (HHHartmann) [inline docs/modules/node.md:669]:
```suggestion
    - `delay_mount` - set to true or false to indicate whether the SPIFFS filesystem mount is delayed until it is first needed or not.
```

File: docs/modules/node.md
Comment (HHHartmann) [inline docs/modules/node.md:667]: Maybe indicate which values are the dafault values.

Comment (HHHartmann) [inline docs/modules/node.md:667]:
```suggestion
    - `banner` - set to true or false to indicate whether the startup banner should be displayed or not. Default: true
```

File: docs/modules/node.md
Comment (HHHartmann) [inline docs/modules/node.md:616]: A reference to user_config.h will be helpful to the unexperienced user.

**Review by HHHartmann — CHANGES_REQUESTED** [review]
Sorry for the late review. the rest LGTM.

File: docs/modules/node.md
Comment (pjsg) [reply docs/modules/node.md]: Done

**Pull request merged**
